Automattic/jetpack
# Search package: run auto config from activation API

#### Changes proposed in this Pull Request:
The PR tries to make the search auto config process idempotent and calls auto config from search activation API. The PR needs `D71473-code` to be functional however it could be shipped without it. Related: #22084. Also fixes #22588.

- Refactored `auto_config_search` functionality to be idempotent and added tests
- Added options NOT to run activation or auto config for the Activation API (changed payload format the activation API consumes, see smoking test section)
- Fixed an issue where `Classic_Search` and `Instant_Search` share the same static `$instance` variable

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Go to a site with Jetpack Search subscription
* Use a non block theme (Block themes are supported thru #22613 )
* Enable Instant Search
* Remove search widgets from sidebars
* Remove overlay sidebar widgets
* Run `jetpack docker wp jetpack-search auto_config username/id`
* Ensure the removed widgets are added again with default filters


* Ensure PHP tests pass: `( cd projects/packages/search && composer test-php )`

##### Smoke test activate plan API:
Note: 
- Remove all widgets in theme sidebar and ovleray sidebar
- All keys in the JSON payload are optional.
- cookie and nonce could be obtained from page `/wp-admin/admin.php?page=jetpack-search`, inspect API requests like `/wp-json/jetpack/v4/search/settings?_cacheBuster=1644280717075`

```
POST https://wordpress.local/?rest_route=/jetpack/v4/search/plan/activate HTTP/1.1
Content-Type: application/json
x-wp-nonce: {{nonce}}
cookie: {{cookie}}

{
        "auto_config_search": false,// default is true,
        "enable_search": true, // default is true,
        "enable_instant_search": true, // default is true,
	"search_plan_info": {
            "search_subscriptions": [{
		"ID": "1",
		"user_id": "1",
		"blog_id": "1",
		"product_id": "2001",
		"expiry": "2117-06-30",
		"subscribed_date": "2017-06-30 16:07:06",
		"renew": true,
		"auto_renew": true,
		"ownership_id": "9698106",
		"most_recent_renew_date": "",
		"subscription_status": "active",
		"product_name": "Jetpack Professional",
		"product_name_en": "Jetpack Professional",
		"product_slug": "jetpack_business",
		"product_type": "bundle",
		"cost": 0,
		"currency": "USD",
		"bill_period": "365",
		"available": "yes",
		"multi": false,
		"support_document": null,
		"is_instant_search": false,
		"tier": null
	}],
	"supports_instant_search": true,
	"supports_only_classic_search": false,
	"supports_search": true,
	"default_upgrade_bill_period": "yearly"
    }
}

```
